### PR TITLE
Check free disk space percent before upgrade

### PR DESCRIPTION
**Pull request by FrankYang0529**
**Problem:**
From [container image lifecycle document](https://kubernetes.io/docs/concepts/architecture/garbage-collection/#container-image-lifecycle), kubelet triggers grabage collection when disk free space percent is higher than `HighThresholdPercent`. This may make some risks in airgapped environment, because it can't get old images if we don't set container registry correctly.

For example, [missing vmdp image issue](https://github.com/harvester/harvester/issues/4534).

**Solution:**
Add a webhook to check free disk space percent like PR [Check disk space before upgrade](https://github.com/harvester/harvester/pull/3700#top).

**Related Issue:**
https://github.com/harvester/harvester/issues/4611

**Test plan:**
### Setup

1. Use this branch to build a new ISO.

### Case 1: a node without enough disk spaces can't upgrade

1. Create a new Harvester cluster. Each node's disk space should be 250G.
2. If every node's `/usr/local` free space is more than 30G, you can use `dd` to write files to make sure one of the node's free disk space is 20G. For example: `dd if=/dev/zero of=/usr/local/test.img bs=1G count=93`.
3. Create a test version.
```yaml
apiVersion: harvesterhci.io/v1beta1
kind: Version
metadata:
  annotations:
    harvesterhci.io/minFreeDiskSpaceGB: "5"
  name: 1.2.2
  namespace: harvester-system
spec:
  isoURL: http://192.168.0.181:8000/harvester-v1.2.2-amd64.iso
  minUpgradableVersion: 1.2.1
  releaseDate: "20231210"
  tags:
  - dev
  - test
```
4. Click upgrade on the dashboard. We should get an error message.

Screenshot 2023-11-03 at 7 41 02 PM

### Case 2: a node with enough disk spaces can upgrade

1. Remove the `dd` file.
2. Click the upgrade again. There is no error message.

**Additional context**
* [X] Rely on https://github.com/harvester/harvester/pull/4658.

### CODE REVIEW

**Review by w13915984028 — COMMENTED** [review]
A question:

would we allow adding an annotation to skip the new check? in none air-gapped cluster, or cluster with image registry, user will have chance to skip such situation, thanks.

**Comment by FrankYang0529** [comment]
> A question:
> 
> would we allow adding an annotation to skip the new check? in none air-gapped cluster, or cluster with image registry, user will have chance to skip such situation, thanks.

Yeah, we already have it.

https://github.com/harvester/harvester/blob/7a7daa7e500eb0d2306fa05106106ccbcb936b10/pkg/webhook/resources/upgrade/validator.go#L126-L130

**Review by w13915984028 — APPROVED** [review]
LGTM, thanks.

**Review by Yu-Jack — APPROVED** [review]
LGTM.

BTW, because I'm curious whether I could register kubelet.config.k8s.io into factory like following, so we might use `Kubelet().V1beta1().KubeletConfiguration().Get()` to get config. 

```
kubeletconfigv1.SchemeGroupVersion.Group: {
	Types: []interface{}{
		kubeletconfigv1.KubeletConfiguration{},
	},
	GenerateTypes:   true,
	GenerateClients: true,
},
```

But, I found it lacks of `KubeletConfigurationList` in types, so it failed.

**Comment by FrankYang0529** [comment]
> BTW, because I'm curious whether I could register kubelet.config.k8s.io into factory like following, so we might use `Kubelet().V1beta1().KubeletConfiguration().Get()` to get config.
> 
> ```
> kubeletconfigv1.SchemeGroupVersion.Group: {
> 	Types: []interface{}{
> 		kubeletconfigv1.KubeletConfiguration{},
> 	},
> 	GenerateTypes:   true,
> 	GenerateClients: true,
> },
> ```
> 
> But, I found it lacks of `KubeletConfigurationList` in types, so it failed.

If I understand correctly, `KubeletConfiguration` is not stored in etcd, so we can't get it from API server. We have to get it from each kubelet server.

File: pkg/webhook/resources/upgrade/validator.go [hunk 1]
```diff
@@ -252,34 +254,33 @@ func (v *upgradeValidator) checkDiskSpace(node *corev1.Node, minFreeDiskSpace ui
...
-	body, err := ioutil.ReadAll(resp.Body)
+	kubeletConfiguration, err := v.getKubeletConfigz(node.Name, kubeletURL)
 	if err != nil {
-		return werror.NewInternalError(fmt.Sprintf("node %s, can't read response from %s, err: %+v", node.Name, kubeletURL, err))
+		return err
 	}
-	if err = json.Unmarshal(body, &summary); err != nil {
-		return werror.NewInternalError(fmt.Sprintf("node %s, can't parse json response %s, err: %+v", node.Name, string(body), err))
+
+	imageGCHighThresholdPercent := 85.0 // default value in kubelet config
```
Comment (Yu-Jack) [inline pkg/webhook/resources/upgrade/validator.go]: BTW, I think we could move default percent to top area as const?

Comment (FrankYang0529) [reply pkg/webhook/resources/upgrade/validator.go]: Add `defaultImageGCHighThresholdPercent`. Thank you.

File: pkg/webhook/resources/upgrade/validator.go [hunk 2]
```diff
@@ -252,34 +255,33 @@ func (v *upgradeValidator) checkDiskSpace(node *corev1.Node, minFreeDiskSpace ui
...
+	imageGCHighThresholdPercent := defaultImageGCHighThresholdPercent
+	if kubeletConfiguration.ImageGCHighThresholdPercent != nil {
+		imageGCHighThresholdPercent = float64(*kubeletConfiguration.ImageGCHighThresholdPercent)
 	}
-	if summary.Node.Fs.AvailableBytes == nil {
-		return werror.NewInternalError(fmt.Sprintf("can't get node %s available bytes from %s, err: %+v", node.Name, kubeletURL, err))
+	usedPercent := (float64(*summary.Node.Fs.UsedBytes) / float64(*summary.Node.Fs.CapacityBytes)) * 100.0
+	logrus.Debugf("node %s uses %3f%% storage space, kubelet image garbage collection threshold is %3f%%", node.Name, usedPercent, imageGCHighThresholdPercent)
+
+	if usedPercent > imageGCHighThresholdPercent {
```
Comment (bk201) [inline pkg/webhook/resources/upgrade/validator.go:281]: @FrankYang0529 I suggest we use the condition `"userPercent + (expected new image capacity)" > imageGCHighThresholdPercent`. The problem we saw before is that after/during preloading images, a GC removes new images. We can add a reasonable value of `(expected new image capacity)` by doing some experiment...

Comment (FrankYang0529) [reply pkg/webhook/resources/upgrade/validator.go:281]: Updated it. Thank you.

**Review by bk201 — APPROVED** [review]
LGTM. The test plan needs an update. QA will need to calculate the a reasonable size which is near the threshold.

**Comment by FrankYang0529** [comment]
@Mergify backport v1.2

**Pull request merged**